### PR TITLE
Websocket extension implementation

### PR DESCRIPTION
**Pull request by arondn2**
This is an implementation of websocket extension.
Features:
- Application base class for ServerIOServer implementation only
- Documentation to implement a websocket server with an existing express instance (like a pure express or RestApplication).
- Decorators to associate controllers with socket namespaces, to associate controller's methods with events and to inject some elements (like socket, message, server or namespace).
- Booter to autoload controllers in the same way that services, repositories and controllers are autoloaded.
- Customizable sequence class to handler events and ways to customize the response (success o fail) for an event.
- Readme documentation.

All this work is based in the next repositories
- https://github.com/raymondfeng/loopback4-example-websocket/
- https://github.com/strongloop/loopback-next/tree/socketio/extensions/socketio

Pending
- Commit example

Signed-off-by: Alexander Rondon <arondn2@gmail.com>

## Checklist

- [x] DCO (Developer Certificate of Origin) [signed in all commits](https://loopback.io/doc/en/contrib/code-contrib.html)
- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: extensions/websocket/README.md [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+    this.wsServer = this.getSync(WebsocketBindings.SERVER);
+  }
+  
+  // Start ws server
+  async start() {
+    await this.wsServer.start();
+  }
+  // Stop ws server
+  async stop() {
+    await this.wsServer.stop();
```
Comment (mdbetancourt) [inline extensions/websocket/README.md:67]: websocket server is not started by default without these lines?

File: extensions/websocket/README.md [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+  }
+  @ws.connect()
+  connect(@ws.socket() socket: Socket) {
+    console.log('Client connected: %s', socket.id);
+  }
+
+  @ws.subscribe('chat message')
+  handleChatMessage(msg: unknown, @ws.socket() socket: Socket) {
+    console.log('Chat message: %s', msg);
+    socket.nsp.emit('chat message', `[${socket.id}] ${msg}`);
```
Comment (mdbetancourt) [inline extensions/websocket/README.md:204]: isn't?
```suggestion
    this.socket.nsp.emit('chat message', `[${socket.id}] ${msg}`);
```

Comment (mdbetancourt) [reply extensions/websocket/README.md:204]: i think should be possible to response just returning a value

**Review by mdbetancourt — COMMENTED** [review]
I think this should live in package folder (rest is there)

File: extensions/websocket/README.md (hunk 2 above)
Comment (arondn2) [reply extensions/websocket/README.md:204]: The case of `handleChatMessage` method wants to show how inject the socket in the method param, but both cases work.

The response to the event is the method return value. That is explained below in the readme. The `socket.nsp.emit` want to show you how emit and event to a namespace or socket.

**Comment by mdbetancourt** [comment]
a question is possible use a plain socket https://nodejs.org/api/net.html#net_class_net_socket would be great i'm working with a iot device and is using plain socket, here already exists another pull request with socket.io #6352 but nothing about plain sockets (or the possibility to change it)

File: extensions/websocket/README.md (hunk 1 above)
Comment (arondn2) [reply extensions/websocket/README.md:67]: to be honest, i am no sure, but, I could swear they are required. I will make some tests

**Comment by arondn2** [comment]
@mdbetancourt I will document myself about it to see what can be done. At the moment, I think it is possible to find examples of how to use socket io for IoT.

**Comment by mdbetancourt** [comment]
> @mdbetancourt I will document myself about it to see what can be done. At the moment, I think it is possible to find examples of how to use socket io for IoT.

yes but this device is closed source, and only allow plain sockets (not websockets) [Stack Overflow](https://stackoverflow.com/questions/4973622/difference-between-socket-and-websocket#:~:text=WebSockets%20typically%20run%20from%20browsers,are%20more%20powerful%20and%20generic.)

**Comment by arondn2** [comment]
> btw, you mixed spanish with english in your pull request message
Thanks. I had fixed.

> yes but this device is closed source, and only allow plain sockets (not websockets) [Stack Overflow]
I think this feature is outside the scope of this extension, but it would be interesting look up an implementation.

**Comment by arondn2** [comment]
@raymondfeng @bajtos @emonddr @agnes512, any sugestion to solved the test "lb4 copyright/updates copyright/license headers with options.exclude:" that is falling in this pull request?

**Comment by achrinza** [comment]
> any suggestion to solve the test "lb4 copyright/updates copyright/license headers

Run `lb4 copyright` in the package's directory

See [these docs](https://loopback.io/doc/en/lb4/code-contrib-lb4.html#create-a-new-package) on how to create a new package.

**Comment by raymondfeng** [comment]
@arondn2 Thank you for the contribution!

Are you aware of a similar effort that is happening on #6352? I think we should consolidate the two PRs if possible.

**Comment by arondn2** [comment]
@raymondfeng I saw that PR yesterday when I started to incorporated this extension, but I still create this PRs to get instructions on what I should to do. Can you tell what a should to do?

**Comment by arondn2** [comment]
> Run `lb4 copyright` in the package's directory

Thanks, I will resolved it

**Comment by raymondfeng** [comment]
> @raymondfeng I saw that PR yesterday when I started to incorporated this extension, but I still create this PRs to get instructions on what I should to do. Can you tell what I should to do?

Is it possible to rebase your PR on top the other one or vice versa? I pinged @mayank-SFIN571 too. They two can have an initial assessment.

**Comment by mayank-SFIN571** [comment]
Hi @arondn2 
The only major difference I can see in [6352](https://github.com/strongloop/loopback-next/pull/6352) as compared to this PR is sequence implementation.
If you feel its alright, then I can implement it there, same as yours, and then we can follow one PR and collaborate together there. You can also do a review of the PR in the current state. It has been reviewed by Raymond and Hage Yappa once. So the review cycle has started there. We may close that one sooner.

**Comment by raymondfeng** [comment]
From the programming model perspective, I would like to explore the possibility to use `async iterator`, as illustrated in http://thecodebarbarian.com/async-generator-functions-and-websockets-in-node-js.html.

For subscribers, they can receive an async iterator so that:

```js
for await (const message of subscription) {
  // process message
}
```

For publishers, they can feed messages into the socket:

```js
for await (const message of publisher) {
  socket.send(message);
}
```

**Comment by arondn2** [comment]
hi, @mayank-SFIN571 
What about if I make a PR to you repo with the changes I made? I think this repo has additionals elements also sequence. Is this ok for you?

**Comment by mayank-SFIN571** [comment]
> hi, @mayank-SFIN571
> What about if I make a PR to you repo with the changes I made? I think this repo has additionals elements also sequence. Is this ok for you?

@arondn2 that sounds great

**Comment by arondn2** [comment]
great @mayank-SFIN571 I'm on it

**Comment by arondn2** [comment]
Closed to work in #6352

**Pull request closed without merging**